packages/x-data-grid-pro/src/components/GridServerSideTreeDataGroupingCell.tsx

```diff
@@ -0,0 +1,134 @@
+import * as React from 'react';
+import { unstable_composeClasses as composeClasses } from '@mui/utils';
+import Box from '@mui/material/Box';
+import Badge from '@mui/material/Badge';
+import {
+  getDataGridUtilityClass,
+  GridRenderCellParams,
+  GridServerSideGroupNode,
+  useGridSelector,
+} from '@mui/x-data-grid';
+import CircularProgress from '@mui/material/CircularProgress';
+import { useGridRootProps } from '../hooks/utils/useGridRootProps';
+import { useGridPrivateApiContext } from '../hooks/utils/useGridPrivateApiContext';
+import { DataGridProProcessedProps } from '../models/dataGridProProps';
+import { GridPrivateApiPro } from '../models/gridApiPro';
+import { GridStatePro } from '../models/gridStatePro';
+
+type OwnerState = DataGridProProcessedProps;
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    root: ['treeDataGroupingCell'],
+    toggle: ['treeDataGroupingCellToggle'],
+    loadingContainer: ['treeDataGroupingCellLoadingContainer'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+interface GridTreeDataGroupingCellProps
+  extends GridRenderCellParams<any, any, any, GridServerSideGroupNode> {
+  hideDescendantCount?: boolean;
+  /**
+   * The cell offset multiplier used for calculating cell offset (`rowNode.depth * offsetMultiplier` px).
+   * @default 2
+   */
+  offsetMultiplier?: number;
+}
+
+interface GridTreeDataGroupingCellIconProps
+  extends Pick<GridTreeDataGroupingCellProps, 'id' | 'field' | 'rowNode' | 'row'> {
+  descendantCount: number;
+}
+
+function GridTreeDataGroupingCellIcon(props: GridTreeDataGroupingCellIconProps) {
+  const apiRef = useGridPrivateApiContext() as React.MutableRefObject<GridPrivateApiPro>;
+  const rootProps = useGridRootProps();
+  const classes = useUtilityClasses(rootProps);
+  const { rowNode, id, field, descendantCount } = props;
+
+  const loadingSelector = (state: GridStatePro) => state.dataSource.loading[id] ?? false;
+  const errorSelector = (state: GridStatePro) => state.dataSource.errors[id];
+  const isDataLoading = useGridSelector(apiRef, loadingSelector);
+  const error = useGridSelector(apiRef, errorSelector);
+
+  const hasServerChildren = rowNode.hasServerChildren;
+
+  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+    if (!rowNode.childrenExpanded) {
+      // always fetch/get from cache the children when the node is expanded
+      apiRef.current.unstable_dataSource.fetchRows(id);
+    } else {
+      apiRef.current.setRowChildrenExpansion(id, !rowNode.childrenExpanded);
```

Comment:
Thinking about this method, should it be calling `apiRef.current.unstable_dataSource.fetchRows(id)` under the hood?

Reply:
Do you mean `unstable_dataSource.fetchRows` should call it internally or the other way around 🤔

Reply:
> is `setRowChildrenExpansion` an atomic operation that only does the expansion?

This is my current understanding, yes.

>  If it's the latter, perhaps we should expose a method that would do the former?

Do you mean a generic method that we use here irrespective of the client side or server side tree data, which calls the `fetchRows` internally if needed and sets the rowChildrenExpansion to true on receiving the rows?

For my knowledge, would this method solve a current problem we have?

Reply:
> Do you mean a generic method that we use here irrespective of the client side or server side tree data, which calls the fetchRows internally if needed and sets the rowChildrenExpansion to true on receiving the rows?

Yes, correct.

> For my knowledge, would this method solve a current problem we have?

Not really a current problem, this could be added later if necessary 👍🏻